gio/gappinfo.c

```diff
@@ -824,12 +846,68 @@ launch_default_with_portal_async (const char          *uri,
   g_free (real_uri);
 }
 
+static void
+launch_default_with_portal_sync (const char          *uri,
+                                 GAppLaunchContext   *context)
+{
+  GDBusConnection *session_bus;
+  GVariantBuilder opt_builder;
+  GVariant *res = NULL;
+  const char *parent_window = NULL;
+  char *real_uri;
+  GError *error = NULL;
+
+  session_bus = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, &error);
+  if (session_bus == NULL)
+    {
+      g_task_report_error (context, NULL, NULL, NULL, error);
+      return;
+    }
+
+  if (context && context->priv->envp)
+    parent_window = g_environ_getenv (context->priv->envp, "PARENT_WINDOW_ID");
+
+  real_uri = real_uri_for_portal (uri, context, NULL, NULL, NULL, &error);
+  if (real_uri == NULL)
+    return;
+
+  g_variant_builder_init (&opt_builder, G_VARIANT_TYPE_VARDICT);
+
+  /* Calling the D-Bus method for the OpenURI portal "protects" the logic from
+   * not ever having the remote method running in case the xdg-desktop-portal
+   * process is not yet running and the caller quits quickly after the call.
+   */
+  res = g_dbus_connection_call_sync (session_bus,
+                                     FLATPAK_OPENURI_PORTAL_BUS_NAME,
+                                     FLATPAK_OPENURI_PORTAL_PATH,
+                                     FLATPAK_OPENURI_PORTAL_IFACE,
+                                     FLATPAK_OPENURI_PORTAL_METHOD,
+                                     g_variant_new ("(ss@a{sv})",
+                                                    parent_window ? parent_window : "",
+                                                    real_uri,
+                                                    g_variant_builder_end (&opt_builder)),
+                                     NULL,
+                                     G_DBUS_CALL_FLAGS_NONE,
+                                     G_MAXINT,
+                                     NULL,
+                                     &error);
+  if (res == NULL)
+    {
+      g_task_report_error (context, NULL, NULL, NULL, error);
+      return;
```

Comment:
Yes, I need to remove the early return and simply let it fall over into the free section at the end of the function.